File: nodes/rust/src/niblit_client.rs
// niblit_client.rs — HTTP client for the Niblit Python REST API (Rust)

use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::time::Duration;

use crate::niblit_state::NiblitStateEnvelope;

#[derive(Debug, Clone)]
pub struct NiblitClient {
    base_url: String,
    api_key: Option<String>,
    client: Client,
}

#[derive(Debug, Deserialize)]
pub struct ChatResponse {
    pub response: String,
    pub session_id: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct HealthResponse {
    pub status: String,
    pub version: Option<String>,
    pub runtime_level: Option<f64>,
}

#[derive(Debug, Serialize)]
struct ChatRequest<'a> {
    message: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    session_id: Option<&'a str>,
}

impl NiblitClient {
    pub fn new(base_url: impl Into<String>, api_key: Option<String>) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(20))
            .build()
            .expect("failed to build HTTP client");

        Self {
            base_url: base_url.into().trim_end_matches('/').to_string(),
            api_key,
            client,
        }
    }

    fn url(&self, path: &str) -> String {
        format!("{}{}", self.base_url, path)
    }

    fn apply_headers(&self, builder: reqwest::RequestBuilder) -> reqwest::RequestBuilder {
        if let Some(key) = &self.api_key {
            builder.header("X-Niblit-Key", key)
        } else {
            builder
        }
    }

    // ── Health ──────────────────────────────────────────────────────────────

    pub async fn health(&self) -> anyhow::Result<HealthResponse> {
        let req = self.apply_headers(self.client.get(self.url("/health")));
        let resp = req.send().await?.json::<HealthResponse>().await?;
        Ok(resp)
    }

    // ── Chat ────────────────────────────────────────────────────────────────

    pub async fn chat(
        &self,
        message: &str,
        session_id: Option<&str>,
    ) -> anyhow::Result<ChatResponse> {
        let body = ChatRequest { message, session_id };
        let req = self
            .apply_headers(self.client.post(self.url("/chat")))
            .json(&body);
        let resp = req.send().await?.json::<ChatResponse>().await?;
        Ok(resp)
    }

    // ── State exchange ──────────────────────────────────────────────────────

    pub async fn push_state(&self, envelope: &NiblitStateEnvelope) -> anyhow::Result<bool> {
        let req = self
            .apply_headers(self.client.post(self.url("/api/state")))
            .json(envelope);
        let status = req.send().await?.status();
        Ok(status.is_success())
    }

    pub async fn pull_state(&self) -> anyhow::Result<Option<NiblitStateEnvelope>> {
        let req = self.apply_headers(self.client.get(self.url("/api/state")));
        let resp = req.send().await?;
        if resp.status().is_success() {
            let env: NiblitStateEnvelope = resp.json().await?;
            if env.verify() {
                Ok(Some(env))
            } else {
                eprintln!("[NiblitClient] State envelope checksum mismatch — ignoring");
                Ok(None)
            }
        } else {
            Ok(None)
        }
    }

    // ── Environment capabilities ────────────────────────────────────────────

    pub async fn report_env_capabilities(
        &self,
        caps: &Value,
    ) -> anyhow::Result<bool> {
        let req = self
            .apply_headers(self.client.post(self.url("/api/env/capabilities")))
            .json(caps);
        let status = req.send().await?.status();
        Ok(status.is_success())
    }
}

File: nodes/typescript/src/niblit-client.ts
/**
 * niblit-client.ts — HTTP client for the Niblit Python REST API
 *
 * Provides a typed interface to Niblit's /chat, /state, and /health endpoints.
 * Handles retries, timeout, and error normalization.
 */

import fetch from "node-fetch";
import { NiblitStateEnvelope, sealEnvelope, verifyEnvelope } from "./niblit-state.js";

export interface NiblitClientConfig {
  baseUrl: string;         // e.g. "http://localhost:8000" or your Fly.io URL
  apiKey?: string;         // Optional NIBLIT_API_KEY header
  timeoutMs?: number;      // Default 15 000
  maxRetries?: number;     // Default 3
}

export interface ChatResponse {
  response: string;
  session_id?: string;
}

export interface HealthResponse {
  status: string;
  version?: string;
  runtime_level?: number;
}

export class NiblitClient {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly timeoutMs: number;
  private readonly maxRetries: number;

  constructor(config: NiblitClientConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, "");
    this.apiKey = config.apiKey;
    this.timeoutMs = config.timeoutMs ?? 15_000;
    this.maxRetries = config.maxRetries ?? 3;
  }

  // ── Chat ──────────────────────────────────────────────────────────────────

  async chat(message: string, sessionId?: string): Promise<ChatResponse> {
    const body: Record<string, string> = { message };
    if (sessionId) body.session_id = sessionId;
    const json = await this._post<ChatResponse>("/chat", body);
    return json;
  }

  // ── State exchange ────────────────────────────────────────────────────────

  /** Push the Node.js state envelope to the Python core. */
  async pushState(envelope: NiblitStateEnvelope): Promise<boolean> {
    const sealed = await sealEnvelope(envelope);
    try {
      await this._post("/api/state", sealed);
      return true;
    } catch {
      return false;
    }
  }

  /** Pull the current state envelope from the Python core. */
  async pullState(): Promise<NiblitStateEnvelope | null> {
    try {
      const data = await this._get<NiblitStateEnvelope>("/api/state");
      const valid = await verifyEnvelope(data);
      if (!valid) {
        console.warn("[NiblitClient] Received state with invalid checksum — ignoring");
        return null;
      }
      return data;
    } catch {
      return null;
    }
  }

  // ── Runtime environment reporting ─────────────────────────────────────────

  /**
   * Report this Node environment's capabilities to the Python core.
   * The core merges these into EnvAdapterRegistry / env_state.
   */
  async reportEnvCapabilities(caps: Record<string, unknown>): Promise<boolean> {
    try {
      await this._post("/api/env/capabilities", {
        runtime: "node",
        platform: `${process.platform}/${process.arch}`,
        node_version: process.version,
        capabilities: caps,
      });
      return true;
    } catch {
      return false;
    }
  }

  // ── Health ────────────────────────────────────────────────────────────────

  async health(): Promise<HealthResponse> {
    return this._get<HealthResponse>("/health");
  }

  // ── Internal ──────────────────────────────────────────────────────────────

  private async _get<T>(path: string): Promise<T> {
    return this._request<T>("GET", path, undefined);
  }

  private async _post<T>(path: string, body: unknown): Promise<T> {
    return this._request<T>("POST", path, body);
  }

  private async _request<T>(
    method: string,
    path: string,
    body: unknown,
    attempt = 1,
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);

    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) headers["X-Niblit-Key"] = this.apiKey;

   try {
  const res = await fetch(url, {
    method,
    headers,
    body: body != null ? JSON.stringify(body) : undefined,
    signal: controller.signal,
  });

      if (!res.ok) {
        const text = await res.text().catch(() => "");
        throw new Error(`HTTP ${res.status}: ${text.slice(0, 200)}`);
      }

      return (await res.json()) as T;
    } catch (err) {
      if (attempt < this.maxRetries) {
        const delay = 200 * 2 ** (attempt - 1);
        await new Promise((r) => setTimeout(r, delay));
        return this._request<T>(method, path, body, attempt + 1);
      }
      throw err;
    } finally {
      clearTimeout(timer);
    }
  }
}
